packager/media/codecs/ec3_audio_util.cc

```diff
@@ -81,10 +81,77 @@ uint8_t ReverseBits8(uint8_t n) {
   return ((n >> 4) & 0x0f) | ((n & 0x0f) << 4);
 }
 
+// Mapping of channel configurations to the MPEG audio value based on
+// ETSI TS 102 366 V1.4.1 Digital Audio Compression (AC-3, Enhanced AC-3)
+// Standard Table I.1.1
+uint32_t EC3ChannelMaptoMPEGValue(uint32_t channel_map) {
+  uint32_t ret = 0;
+
+  switch (channel_map) {
+    case kCenter:
+      ret = 1;
+      break;
+    case kLeft | kRight:
+      ret = 2;
+      break;
+    case kCenter| kLeft | kRight:
+      ret = 3;
+      break;
+    case kCenter | kLeft | kRight | kCenterSurround:
+      ret = 4;
+      break;
+    case kCenter | kLeft | kRight | kLeftSurround | kRightSurround:
+      ret = 5;
+      break;
+    case kCenter | kLeft | kRight | kLeftSurround | kRightSurround |
+         kLFEScreen:
+      ret = 6;
+      break;
+    case kCenter | kLeft | kRight | kLwRwPair | kLeftSurround | kRightSurround |
+         kLFEScreen:
+      ret = 7;
+      break;
+    case kLeft | kRight | kCenterSurround:
+      ret = 9;
+      break;
+    case kLeft | kRight | kLeftSurround | kRightSurround:
+      ret = 10;
+      break;
+    case kCenter | kLeft | kRight | kLrsRrsPair | kCenterSurround | kLFEScreen:
+      ret = 11;
+      break;
+    case kCenter | kLeft | kRight | kLeftSurround | kRightSurround |
+         kLrsRrsPair | kLFEScreen:
+      ret = 12;
+      break;
+    case kCenter | kLeft | kRight | kLeftSurround | kRightSurround |
+         kLFEScreen | kLvhRvhPair:
+      ret = 14;
+      break;
+    case kCenter | kLeft | kRight | kLeftSurround | kRightSurround |
+         kLFEScreen | kLvhRvhPair | kLtsRtsPair:
+      ret = 16;
+      break;
+    case kCenter | kLeft | kRight | kLeftSurround | kRightSurround |
+         kLFEScreen | kLvhRvhPair | kCenterVerticalHeight | kLtsRtsPair |
+         kTopCenterSurround:
+      ret = 17;
+      break;
+    case kCenter | kLeft | kRight | kLsdRsdPair | kLrsRrsPair | kLFEScreen |
+         kLvhRvhPair | kLtsRtsPair:
+      ret = 19;
+      break;
+    default:
+      ret = 0xFFFFFFFF;
```

Comment:
Should we consider it as an error or will it happen in practice that a channel map does not have a MPEG value?

Reply:
All the “normal” ones can be converted. However, if find one that cannot be converted, I don’t think the spec says it is invalid, just unusual. In such situation, Dolby's channel scheme shall be used.